src/FreeAge/client/render_window.cpp

```diff
@@ -3296,6 +3407,32 @@ void RenderWindow::DequeueProductionQueueItem(int queueIndex) {
   connection->Write(CreateDequeueProductionQueueItemMessage(selection.front(), queueIndexFromBack));
 }
 
+void RenderWindow::UngarrisonUnit(int garrisonedUnitIndex) {
+  if (selection.size() != 1) {
+    LOG(ERROR) << "Can only ungarrison a unit if only a single building or unit is selected.";
+    return;
+  }
+  
+  u32 objectId = selection.front();
+  auto objectIt = map->GetObjects().find(objectId);
+  if (objectIt == map->GetObjects().end()) {
+    LOG(ERROR) << "Failed to find the id of the first object of the selection.";
+    return;
+  }
+  ClientUnit* garrisonedUnit = objectIt->second->GetGarrisonedUnits().at(garrisonedUnitIndex);
+  // NOTE: Could be simplified if Objects stored there ID. #ids
```

Comment:
I think having IDs in the object is a must... Basically ids are needed for the messages and the pointers for everything else. Now except for the having to pass id+pointer to functions (no only in this pr, and in some previous code), you also have inconsistency in what it is stored. For example the only places that there are ids instead of pointers is the player's selection of objects, and that because is mostly used to create messages, but in the future you would also need the pointers to draw the selected units in the interface.

Reply:
I agree that it would be very good to simplify this. Perhaps almost the same level of convenience as with the ID storage in the object could also be achieved with some kind of helper class `ObjectPtr` that stores the ID, the pointer, and possibly overloads `operator ->` to act like a direct pointer (while the ID could be retrieved with `.id()`). Kind of like a smart pointer, but for keeping the ID instead of for being smart. (Just as another idea; putting the IDs into the objects is also fine.)